pkg/v1/remote/transport/bearer.go

```diff
@@ -67,24 +67,54 @@ func stringSet(ss []string) map[string]struct{} {
 
 // RoundTrip implements http.RoundTripper
 func (bt *bearerTransport) RoundTrip(in *http.Request) (*http.Response, error) {
-	sendRequest := func() (*http.Response, error) {
+	sendRequest := func(registryToken string) (*http.Response, error) {
 		// http.Client handles redirects at a layer above the http.RoundTripper
 		// abstraction, so to avoid forwarding Authorization headers to places
 		// we are redirected, only set it when the authorization header matches
 		// the registry with which we are interacting.
 		// In case of redirect http.Client can use an empty Host, check URL too.
 		if matchesHost(bt.registry, in, bt.scheme) {
-			hdr := fmt.Sprintf("Bearer %s", bt.bearer.RegistryToken)
+			hdr := fmt.Sprintf("Bearer %s", registryToken)
 			in.Header.Set("Authorization", hdr)
 		}
 		return bt.inner.RoundTrip(in)
 	}
 
-	res, err := sendRequest()
-	if err != nil {
-		return nil, err
+	var res *http.Response
+	var refreshed bool
+	var err error
+	if len(bt.bearer) == 0 {
+		// If bt.bearer is empty, means the token is not initialed, should refresh it firstly
+		if err := bt.refresh(in.Context()); err != nil {
+			return nil, err
+		}
 	}
+	for index := range bt.bearer {
+		res, err = sendRequest(bt.bearer[index].RegistryToken)
+		if err != nil {
+			return nil, err
+		}
 
+		if refreshed, err = bt.refreshWithChallenges(in, res); err != nil {
+			return nil, err
+		}
+		if refreshed {
+			res, err = sendRequest(bt.bearer[index].RegistryToken)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		if res.StatusCode == http.StatusOK {
+			return res, nil
+		}
```

Comment:
I don't think this is quite right. We want to exit the loop early if the request was successful, but "successful" isn't necessarily a 200, right? We only want to keep going through the loop when the response still is a 401 with challenges.

Reply:
Fixed

Reply:
cc @jonjohnsonjr